ctapipe/image/cleaning.py

```diff
@@ -241,6 +243,56 @@ def apply_time_delta_cleaning(
     return pixels_to_keep
 
 
+def apply_time_average_cleaning(
+    geom, mask, image, arrival_times, picture_thresh, time_limit
+):
+    """
+    Extract all pixels that arrived within a given timeframe 
+    with respect to the time average of the pixels on the main island.
+
+    Parameters
+    ----------
+    geom: `ctapipe.instrument.CameraGeometry`
+        Camera geometry information
+    mask: array, boolean
+        boolean mask of *clean* pixels before time_delta_cleaning
+    image: array
+        pixel values
+    arrival_times: array
+        pixel timing information
+    picture_thresh: float
+        threshold above which time limit is extended twice its value
+    time_limit: int or float
+        arrival time limit w.r.t. the average time of the core pixels
+
+    Returns
+    -------
+
+    A boolean mask of *clean* pixels.  To get a zero-suppressed image and pixel
+    list, use `image[mask], geom.pix_id[mask]`, or to keep the same
+    image size and just set unclean pixels to 0 or similar, use
+    `image[~mask] = 0`
+
+    """
+    pixels_to_remove = []
+    mask = mask.copy()
+    if sum(mask) > 0:
+
+        # use main island (maximum charge) for time average calculation
+        num_islands, island_labels = number_of_islands(geom, mask)
+        mask_main = island_labels == np.argmax([np.sum(image[np.where(island_labels == l)]) for l in range(1,num_islands+1)]) + 1
+        time_ave = np.average(arrival_times[np.where(mask_main)[0]], weights=image[np.where(mask_main)[0]]**2)
+
+        for pixel in np.where(mask)[0]:
+            time_diff = np.abs(arrival_times[pixel] - time_ave)
+            time_limit = time_limit if image[pixel] < picture_thresh * 2 else time_limit * 2
```

Comment:
This overwrites time_limit. So for each core pixel, you will increase the time limit by a factor of two of all following pixels. This seems not intended.

Also, what is the intension here? Why are brighter pixels allowed a more lax time requirement? Why the split at 2 * picture_threshold?

Reply:
This is a way to avoid removing signal pixels located at edges of elongated shower images.
Images with large impact parameters have large time gradient, resulting in large time difference even inside core pixels. If we use time difference from the average, we have to take care of those events.
Of course, the picture_threshold *2, or time_limit *2 can be changed. This is just a fixed number used in MAGIC.

Reply:
If you have any suggestions, please let me know. This part can be considered further.